Allow globbing in add() and update() AND fix defunct reset(0)

Finii opened the pull request:
**[why]**

We can only add specific files are all files.
It is not possible to add a whole subdirectory via glob.

With add() we can not add removed files. For some unknown reason add()
is public but update() is not.

**[how]**

Forward glob string to libgit2 add() function.

Make update() public.
Allow globs to be specified for update(), to allow updating whole
subdirectories and not only the complete repository.

Finii committed:
GitRepository: Allow to specify glob with add()

[why]
We can only add specific files are all files.
It is not possible to add a whole subdirectory via glob.

[how]
Forward glob string to libgit2 function.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
GitRepository: Make update() public and globby

[why]
With add() we can not add removed files. For some unknown reason add()
is public but update() is not.

[how]
Make update() public.
Allow globs to be specified for update(), to allow updating whole
subdirectories and not only the complete repository.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

alt-graph commented on include/libgit4cpp/GitRepository.h:
```diff
@@ -79,7 +79,7 @@ class GitRepository
 
     /// Stage all new and changed files and folders in the repository directory.
-    void add();
+    void add(const std::string& glob = "*");
```
I think the "glob" parameter needs some explanation. Can I set it to "*.txt" to stage all new/changed text files instead of all?

alt-graph replied:
... and may `glob` be empty?

Finii replied:
>  Can I set it to "*.txt"

I don't know, probably. Setting it is rather easy ;-D

> stage all new/changed text files instead of all

That's how it is supposed to work, although I only tested globs ending in a star.
The upstream documentation is ... nonexisting (or I am too dump).

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/8a0e7dfd-3e94-4f89-ae59-4dc176478fb5)

Without reverse-engineering I can not say more than the few words given above.

Finii replied:
Added docu and a lot of tests that show hopefully everything.

![image](https://github.com/taskolib/libgit4cpp/assets/16012374/a8980011-7d5a-4c4c-8e5a-cc92a460e70f)

alt-graph commented on include/libgit4cpp/GitRepository.h:
```diff
@@ -81,6 +81,12 @@ class GitRepository
+     * Update the tracked files in the repository.
+     * This member function stages all changes of already tracked files in the repository.
+     */
+    void update(const std::string& glob = "*");
```
Same here. The `glob` parameter needs some explanation.

alt-graph reviewed (changes requested):
I also think the globbing `GitRepository::add(...)` should have at least one unit test which shows that and how globbing works.

Finii commented:
I have a refactor PR for this pending, and additionally I now have problems using git reset... sigh.
Rewrote the other lib's commit stuff. Oh my!

Ah, and I want dump-to-stream ops, that comes next. debugging blindly or writing always anew an output function :roll_eyes:

Finii commented:
> additionally I now have problems using git reset

```console
test_GitRepository.cc:     * TODO: Fuktionalitaet von git reset implementieren
```

:sob:

Finii committed:
GitRepository: Fix reset(0) i.e. for reset to HEAD

[why]
It tries to reset to 'HEAD~0' which would in principle be fine, except
that it does not always pick HEAD (the reason for that is unclear for
me).

[how]
Add special rule that 'HEAD' is used as reference if we do not want to
go back but just reset to HEAD.

Prevent negative numbers as they do not make sense and we can avoid any
checks by using unsigneds.

[note]
In principle the interface with a number might be convenient at times
but is very limiting (i.e. unusable) generally. Better would be at least
an additional interface with a free reference string like done for
get_commit().

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
Hmm, repaired `reset()`.

Which is unrelated to the initial topic of this PR.

I would really just like to push in the corrections, I can work on the documentation and tests later.
But at the moment I struggle with `taskolib` and we need the fixes to troddle along there.

I would propose: If these changes are good enough, just doc and tests missing, we could pull them already and I work on the docs tomorrow or something. Well, after `taskolib` that is. But I guess that is almost finished now.

What do you think?

And as I said before, I have lots of commits for this repo flying around anyhow, that could be a nice next PR here.

_Edit:_

I also find the referencing with numbers (example 4) as `HEAD~4` very biased. What is with `HEAD^4`? Is that no use-case?

Finii committed:
tests: GitRepository: Remove reference to "Sequences"

[why]
Although this has been intended to be used to store sequences in the
taskolib the library itself should be generic.

[how]
Just rename the main test working directory.
Thereby make it a constant instead of a lot individual string literals.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
GitRepository: Make vector<FileStatus> a distinct type

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
GitRepository: Fix defunct get_commit(n)

[why]
Getting the n-th anchestor of HEAD does not work, it always complains
that the ref can not be found.

[how]
Use libgit2's specific function to do this.

Also make get_commit()'s number argument unsigned to have no problems
with negative numbers.

Also drop get_commit(void) and make that special / default case of
get_commit(unsigned).

Drop special handling in reset() which is not needed anymore.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
test: GitRepository: Add tests for reset()

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii committed:
tests: GitRepository: Add tests for add() with glob

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
> I also think the globbing GitRepository::add(...) should have at least one unit test which shows that and how globbing works.

Added test.
Fixed `reset(n)`.
Docs coming soon (after baking Christmas cookies :-} ).

Finii committed:
GitRepository: Remove \n from exception messages

[why]
Exception messages should normally not end with "\n".
Our typical style also has no fullstop at the end.

Here all varieties are used.

[how]
Unify.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
There was a conflict with `main`, fixed and rebase -> force push.

Finii commented:
Rebase on master after that merged the refactoring; force push.

Finii commented:
In the output operator I called the repo state `RepoState`.
The type that holds it is named thus: `using RepoState = std::vector<FileStatus>;`

The function to get the repo state is called `status()` which is kind of a misnomer because that is the repo state you get back and not the repo status. Probably we should rename the function too. I did not do that yet to raise this discussion.

Also `collect_status()` should be `collect_state()` or `collect_statuses()`.

_note_
I also considered making `RepoState` a true type and not just an alias, which does have some benefits, but as it is in fact just a `std::vector` and inheriting that (to get a nice own type with empty base optimization) that would not be advisable (see Core Guidelines) and the vector as a data member would need accessors and such, and the handling will become much more convoluted.

Finii committed:
GitRepository: Add more docs about the globbing

with extensive examples in the tests.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>

Finii commented:
Sorry, again force push, reordered the tests to make more sens, building from simple to complex.

alt-graph reviewed (approved):
:+1:

Finii commented:
Solving merge conflict after pulling the other PR into main, force push

Finii commented on tests/test_GitRepository.cc:
```diff
@@ -28,43 +28,43 @@
- *      $name$/   \n
- *          file0.txt   << $msg$ /n file0  \n
- *          file1.txt   << $msg$ /n file1  \n
+ * Filestructure:
```
I guess this needs a code block

Finii committed:
test: Fix comments

Fix whitespaces:
Part of one comment has been created via `tree` but that seems to create
not only blanks (ascii 20) but also something else (ascii a0).
Convert the a0 to 20.

And one part of the comment looks like Doxygen but contains preformatted
code. Although the tests are currently not part of the Doxygen-ization
process we should not have broken output once it is.

Signed-off-by: Fini Jastrow <ulf.fini.jastrow@desy.de>
